src/VisualStudio/Core/Def/DocumentOutline/DocumentOutlineControl.xaml.cs

```diff
@@ -0,0 +1,388 @@
+    {
+        public IVsEditorAdaptersFactoryService EditorAdaptersFactoryService { get; }
+        public IVsCodeWindow CodeWindow { get; }
+
+        private IThreadingContext ThreadingContext { get; }
+        private ILanguageServiceBrokerShim LanguageServiceBroker { get; }
+
+        /// <summary>
+        /// The type of sorting applied to the document model.
+        /// </summary>
+        private SortOption SortOption { get; set; }
+
+        /// <summary>
+        /// Queue to batch up work to do to get the current document model. Used so we can batch up a lot of events 
+        /// and only fetch the model once for every batch.
+        /// </summary>
+        private readonly AsyncBatchingWorkQueue _getModelQueue;
+
+        /// <summary>
+        /// Queue to batch up work to do to update the current document model and UI.
+        /// </summary>
+        private readonly AsyncBatchingWorkQueue _updateModelQueue;
+
+        /// <summary>
+        /// Queue to batch up work to do to highlight the currently selected symbol node.
+        /// </summary>
+        private readonly AsyncBatchingWorkQueue _highlightNodeQueue;
+
+        /// <summary>
+        /// The text snapshot from when the document symbol request was made.
+        /// </summary>
+        private ITextSnapshot? LspSnapshot { get; set; }
+
+        /// <summary>
+        /// Keeps track of the current primary and secondary text views.
+        /// </summary>
+        private readonly Dictionary<IVsTextView, ITextView> _trackedTextViews = new();
+
+        /// <summary>
+        /// Stores the latest document model returned by GetModelAsync to be used by UpdateModelAsync.
+        /// </summary>
+        private ImmutableArray<DocumentSymbolViewModel> LatestFetchedDocumentModel { get; set; }
+
+        /// <summary>
+        /// Is true when LatestFetchedDocumentModel is not empty.
+        /// </summary>
+        [MemberNotNullWhen(true, nameof(LspSnapshot))]
+        private bool LatestFetchedDocumentModelIsNotEmpty { get; set; }
+
+        public DocumentOutlineControl(
+            ILanguageServiceBrokerShim languageServiceBroker,
+            IThreadingContext threadingContext,
+            IAsynchronousOperationListener asyncListener,
+            IVsEditorAdaptersFactoryService editorAdaptersFactoryService,
+            IVsCodeWindow codeWindow)
+        {
+            InitializeComponent();
+
+            ThreadingContext = threadingContext;
+            LanguageServiceBroker = languageServiceBroker;
+            EditorAdaptersFactoryService = editorAdaptersFactoryService;
+            CodeWindow = codeWindow;
+            ComEventSink.Advise<IVsCodeWindowEvents>(codeWindow, this);
+            SortOption = SortOption.Order;
+
+            _getModelQueue = new AsyncBatchingWorkQueue(
+                    DelayTimeSpan.Short,
+                    GetModelAsync,
+                    asyncListener,
+                    threadingContext.DisposalToken);
```

Comment:
It seems we should have a cancellation token to use that's better than threadingContext.DisposalToken here -- otherwise if I close the document entirely all of this background work could still be running, and end up also touching views and such that have since been closed.